pandas/core/indexes/datetimes.py

```diff
@@ -276,35 +297,43 @@ def _simple_new(cls, values, name=None, freq=None, tz=None, dtype=None):
         if we are passed a non-dtype compat, then coerce using the constructor
         """
         # DatetimeArray._simple_new will accept either i8 or M8[ns] dtypes
-        assert isinstance(values, np.ndarray), type(values)
+        values = DatetimeArray._simple_new(values, freq=freq, tz=tz)
```

Comment:
Is it needed to still pass this through to `DatetimeArray` constructor? We cannot yet assume `values` here is a DatetimeArray?

Reply:
I don't know, but I could easily see us getting `i8` values here.

Reply:
This seems possible with a few changes...

1. Change DatetimeIndex.__new__ to pass the DatetimeArray, rather than unboxing the data (this is a good idea anyway)
2. Index._shallow_copy does a `values = getattr(values, '_values', values)`. I don't know if that's for compat with DTI or other index classes, but simply removing it works for DTI and freinds
3. Everywhere calling Datetime(like)Index._shallow_copy with an ndarray would need to be updated.
4. A scattering of other places (e.g. DatetimeIndex.sort_values)